File: worldmodel.py
import entities
import pygame
import ordered_list
import actions
import occ_grid
import point

BLOB_RATE_SCALE = 4
BLOB_ANIMATION_RATE_SCALE = 50
BLOB_ANIMATION_MIN = 1
BLOB_ANIMATION_MAX = 3

ORE_CORRUPT_MIN = 20000
ORE_CORRUPT_MAX = 30000

QUAKE_STEPS = 10
QUAKE_DURATION = 1100
QUAKE_ANIMATION_RATE = 100

VEIN_SPAWN_DELAY = 500
VEIN_RATE_MIN = 8000
VEIN_RATE_MAX = 17000

class WorldModel:
    def __init__(self, num_rows, num_cols, background):
      self.background = occ_grid.Grid(num_cols, num_rows, background)
      self.num_rows = num_rows
      self.num_cols = num_cols
      self.occupancy = occ_grid.Grid(num_cols, num_rows, None)
      self.entities = []
      self.action_queue = ordered_list.OrderedList()
      
    def add_entity(self, entity):
       pt = entity.get_position()
       if self.within_bounds(pt):
          old_entity = occ_grid.get_cell(self.occupancy, pt)
          if old_entity != None:
             old_entity.clear_pending_actions()
          occ_grid.set_cell(self.occupancy, pt, entity)
          self.entities.append(entity)


    def within_bounds(self, pt):
       return (pt.x >= 0 and pt.x < self.num_cols and
          pt.y >= 0 and pt.y < self.num_rows)


    def is_occupied(self, pt): 
       return (self.within_bounds(pt) and
          occ_grid.get_cell(self.occupancy, pt) != None)
          
    def find_nearest(self, pt, type):
       oftype = [(e, pt.distance_sq(e.get_position()))
          for e in self.entities if isinstance(e, type)]

       return nearest_entity(oftype)
       
    def move_entity(self, entity, pt):
        tiles = []
        if self.within_bounds(pt):
            old_pt = entity.get_position()
            occ_grid.set_cell(self.occupancy, old_pt, None)
            tiles.append(old_pt)
            occ_grid.set_cell(self.occupancy, pt, entity)
            tiles.append(pt)
            entity.set_position(pt)

        return tiles
        
    def remove_entity(self, entity):
        self.remove_entity_at(entity.get_position())
        
    def remove_entity_at(self, pt):
        if (self.within_bounds(pt) and
            occ_grid.get_cell(self.occupancy, pt) != None):
            entity = occ_grid.get_cell(self.occupancy, pt)
            entity.set_position(point.Point(-1, -1))
            self.entities.remove(entity)
            occ_grid.set_cell(self.occupancy, pt, None)
        
    def schedule_action(self, action, time):
        self.action_queue.insert(action, time)


    def unschedule_action(self, action):
        self.action_queue.remove(action)
        
        
        
    def update_on_time(self, ticks):
        tiles = []

        next = self.action_queue.head()
        while next and next.ord < ticks:
            self.action_queue.pop()
            tiles.extend(next.item(ticks))  # invoke action function
            next = self.action_queue.head()

        return tiles
        
        
    def get_background_image(self, pt): #worldview
        if self.within_bounds(pt):
            return occ_grid.get_cell(self.background, pt).get_image()


    def get_background(self, pt):
        if self.within_bounds(self, pt):
            return occ_grid.get_cell(self.background, pt)


    def set_background(self, pt, bgnd):
        if self.within_bounds(pt):
            occ_grid.set_cell(self.background, pt, bgnd)
                
    def get_tile_occupant(self, pt): #worldview
        if self.within_bounds(pt):
            return occ_grid.get_cell(self.occupancy, pt)

    def get_entities(self):
        return self.entities
        
    def create_vein(self, name, pt, ticks, i_store):
        vein = entities.Vein("vein" + name,
        random.randint(VEIN_RATE_MIN, VEIN_RATE_MAX),
        pt, image_store.get_images(i_store, 'vein'))
        return vein
        
    def clear_pending_actions(self, entity):
       for action in entity.get_pending_actions():
          self.unschedule_action(action)
       entity.clear_pending_actions()  
    
    def find_open_around(self, pt, distance):
       for dy in range(-distance, distance + 1):
          for dx in range(-distance, distance + 1):
             new_pt = point.Point(pt.x + dx, pt.y + dy)

             if (self.within_bounds(new_pt) and
                (not self.is_occupied(new_pt))):
                return new_pt

       return None
       
    def create_ore(self, name, pt, ticks, i_store):
        ore = entities.Ore(name, pt, image_store.get_images(i_store, 'ore'),
        random.randint(ORE_CORRUPT_MIN, ORE_CORRUPT_MAX))
        ore.schedule_ore(self,ticks, i_store)

        return ore
    
    def create_blob(self, name, pt, rate, ticks, i_store):
       blob = entities.OreBlob(name, pt, rate,
          image_store.get_images(i_store, 'blob'),
          random.randint(BLOB_ANIMATION_MIN, BLOB_ANIMATION_MAX)
          * BLOB_ANIMATION_RATE_SCALE)
       blob.schedule_blob(self,ticks, i_store)
       return blob
       
    def create_quake(self, pt, ticks, i_store):
       quake = entities.Quake("quake", pt,
          image_store.get_images(i_store, 'quake'), QUAKE_ANIMATION_RATE)
       quake.schedule_quake(self, ticks)
       return quake
       
       
    def blob_next_position(self, entity_pt, dest_pt):
       horiz = sign(dest_pt.x - entity_pt.x)
       new_pt = point.Point(entity_pt.x + horiz, entity_pt.y)

       if horiz == 0 or (self.is_occupied(new_pt) and
          not isinstance(self.get_tile_occupant(new_pt),
          entities.Ore)):
          vert = sign(dest_pt.y - entity_pt.y)
          new_pt = point.Point(entity_pt.x, entity_pt.y + vert)

          if vert == 0 or (self.is_occupied(new_pt) and
             not isinstance(self.get_tile_occupant(new_pt),
             entities.Ore)):
             new_pt = point.Point(entity_pt.x, entity_pt.y)

       return new_pt
        
   
   
def nearest_entity(entity_dists):
   if len(entity_dists) > 0:
      pair = entity_dists[0]
      for other in entity_dists:
         if other[1] < pair[1]:
            pair = other
      nearest = pair[0]
   else:
      nearest = None

   return nearest

File: Java/WorldModel.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Random;
import java.util.function.LongConsumer;

import processing.core.*;

public class WorldModel {
	public static final int BLOB_RATE_SCALE = 4;
	public static final int	BLOB_ANIMATION_RATE_SCALE = 50;
	public static final int	BLOB_ANIMATION_MIN = 1;
	public static final int	BLOB_ANIMATION_MAX = 3;

	public static final int	ORE_CORRUPT_MIN = 20000;
	public static final int	ORE_CORRUPT_MAX = 30000;

	public static final int	QUAKE_STEPS = 10;
	public static final int	QUAKE_DURATION = 1100;
	public static final int	QUAKE_ANIMATION_RATE = 100;
	public static final int VEIN_SPAWN_DELAY = 500;
	public static final int	VEIN_RATE_MIN = 8000;
	public static final int	VEIN_RATE_MAX = 17000;
	
	private int num_rows, num_cols;
	private OCCGrid background, occupancy;
	private List<Entities> entity;
	public OrderedList<LongConsumer> actionQueue;
	public WorldModel(int num_rows, int num_cols)
	{
		this.num_cols = num_cols;
		this.num_rows = num_rows;
		this.background = new OCCGrid(100,100, null);
		this.occupancy = new OCCGrid(100,100, null);
	}
	public WorldModel(int num_cols, int num_rows, Entities background)
	{

		//System.out.println("rows: " + num_rows);
		//System.out.println("cols: " + num_cols);
		this.num_rows = num_rows;
		this.num_cols = num_cols;
		this.background = new OCCGrid(num_cols, num_rows, background);
		this.occupancy = new OCCGrid(num_cols, num_rows, null);
		this.entity = new ArrayList<Entities>();
		this.actionQueue = new OrderedList<LongConsumer>();
	}
	
	public void addEntity(Entities e)
	{
		Point pt = e.getPosition();
		if(withinBounds(pt))
		{
			Actions oldEntity = (Actions) this.occupancy.getCell(pt);
			if(oldEntity != null)
			{
				oldEntity.clearPendingActions();
			}
		}
		this.occupancy.setCell(pt, e);
		this.entity.add(e);
		
	}
	
	public boolean withinBounds(Point pt)
	{
		/***
		 * Checks to see if a point is in the world
		 * @param: a point
		 * @return: True if point is in the world; if otherwise, false
		 */
		boolean xBounds = (pt.getX() >= 0) && (pt.getX() < this.num_cols);
		boolean yBounds = (pt.getY() >= 0) && (pt.getY() < this.num_rows);
		return xBounds && yBounds; 
	}
	
	public boolean isOccupied(Point pt)
	{
		return withinBounds(pt) && (this.occupancy.getCell(pt) != null);
	}
	
	public Entities findNearest(Point pt, Class type)
	{
		/***
		 * returns index of the ones of the same type w/ nearest
		 * @param pt, a point in the world
		 * @param e, the type of entity you're checking
		 * @return int array of the indexes of the checked Entity type
		 */
		ArrayList<Integer> same = new ArrayList<Integer>();
		for(int i = 0; i < this.entity.size(); i++)
		{
			if( this.entity.get(i).getClass() == type)
			{
				same.add(i);
			}
		}
		return nearestEntity(same, pt);
	}
	
	public List<Point> moveEntity(Entities e, Point pt)
	{
		/***
		 * moves entity to a new point
		 * @param e, an entity
		 * @param pt, a point
		 * @return the points where the entity was and will be
		 */
		ArrayList<Point> tiles = new ArrayList<Point>();
		if (withinBounds(pt))
		{
			Point oldPoint = e.getPosition();
			this.occupancy.setCell(oldPoint, null);
			tiles.add(oldPoint);
			this.occupancy.setCell(pt, e);
			tiles.add(pt);
			e.setPosition(pt);
		}
		return tiles;
	}
	
	public void removeEntity(Entities e)
	{
		removeEntityAt(e.getPosition());
	}
	
	public void removeEntityAt(Point pt)
	{
		if(isOccupied(pt))
		{
			Entities e = this.occupancy.getCell(pt);
			e.setPosition(new Point(-1,-1));
			this.entity.remove(this.entity.indexOf(e));
			this.occupancy.setCell(pt, null);
		}
	}
	
	public void scheduleAction(LongConsumer action, long time)
	{
		actionQueue.insert(action, time);
		System.out.println(actionQueue.size());
	}
	
	public void unscheduleAction(LongConsumer action)
	{
		actionQueue.remove(action);
	}
	
	public void updateOnTime(long ticks)
	{
		OrderedList.ListItem<LongConsumer> next = actionQueue.head();
		
		while(next != null && next.ord < ticks)
		{
			actionQueue.pop();
			next.item.accept(ticks);
			System.out.println("here");
			next = actionQueue.head();
		}
		

	}
	
	public PImage getBackgroundImage(Point pt)
	{
		if(withinBounds(pt))
		{
			return this.background.getCell(pt).getImage();
		}
		return null; 
	}
	
	public Entities getBackground(Point pt)
	{
		/***
		 * determines which entity is at a point
		 * @param pt, a point
		 * @return the entity at pt, otherwise null
		 */
		if(withinBounds(pt))
		{
			return this.background.getCell(pt);
		}
		
		return null;
	}
	
	public void setBackground(Point pt, Entities b)
	{
		/***
		 * puts an entity at a point
		 * @param pt, a point
		 * @param b, an entity
		 * sets the cell (specified from pt) to a specific entity (b)
		 */
		if(withinBounds(pt))
		{
			this.background.setCell(pt,b);
		}
	}
	
	public Entities getTileOccupant(Point pt)
	{
		/***
		 * determines which entity is currently at a point
		 * @param pt, a point
		 * @return if the point is on the screen, return the entity that is at that point
		 * otherwise: null
		 */
		if (withinBounds(pt))
		{
			return this.occupancy.getCell(pt);
		}
		return null; 
	}
	
	
	public List<Entities> getEntities()
	{
		return this.entity;
	}
	
	public Point nextPosition(Point ePt, Point destPt)
	{
		/***
		 * determines where an entity is going to move next 
		 * @param ePt, point with an entity
		 * @param destPt, point where the entity will go
		 * @return the point where the entity will move next
		 */
		int horz = Entities.sign((int)(destPt.getX() - ePt.getY()));
		Point newPoint = new Point(ePt.getX() + horz, ePt.getY());
		
		if( horz == 0 || isOccupied(newPoint))
		{
			int vert = Entities.sign((int)(destPt.getY() - ePt.getY()));
			newPoint = new Point(ePt.getX(), ePt.getY() + vert);
			
			if(vert == 0 || isOccupied(newPoint))
			{
				newPoint = new Point(ePt.getX(), ePt.getY());
			}
		}
		return newPoint;
	}
	
	public Vein createVein(String name, Point pt, long ticks, HashMap<String, ArrayList<PImage>> iStore)
	{
		Random rand = new Random();
		int randInt = rand.nextInt(this.VEIN_RATE_MAX- this.VEIN_RATE_MIN) + this.VEIN_RATE_MIN;
		
		Vein vein = new Vein("vein" + name, pt, ImageStore.getImages(iStore, "vein"), randInt);
		return vein;
	}
	
	public void clearPendingActions(Actions e)
	{
		for(LongConsumer action: e.getPendingActions())
		{
			unscheduleAction(action);
		}
		e.clearPendingActions();
	}
	
	public Point findOpenAround(Point pt, double distance)
	{
		/***
		 * checks if the area around is point is open
		 * @param pt, a point
		 * @param distance, the distance around the point
		 * @return an open point if it's on the screen & isn't occupied, otherwise, null
		 */
		for (int i = (int)(-distance); i < (int)(distance+1); i++)
		{
			for(int j = (int)(-distance); j < (int)(distance+1); j++)
			{
				Point newPoint = new Point((pt.getX() + i),(pt.getY()+ j));
				if(withinBounds(newPoint) && !isOccupied(newPoint))
				{
					return newPoint;
				}
			}
		}
		return null;
	}	
	
	public Ore createOre(String name, Point pt, long current_ticks, HashMap<String, ArrayList<PImage>> iStore)
	{
		Random rand = new Random();
		int temp = rand.nextInt(ORE_CORRUPT_MAX - ORE_CORRUPT_MIN) + ORE_CORRUPT_MIN;
		Ore ore = new Ore(name, pt, ImageStore.getImages(iStore, "ore"),temp);
		ore.scheduleOre(this,current_ticks, iStore);
		
		return ore;
	}
	
	public OreBlob createBlob(String name, Point pt, long rate, long ticks, HashMap<String, ArrayList<PImage>> i_store)
	{
		Random rand = new Random();
		int temp = rand.nextInt(BLOB_ANIMATION_MAX - BLOB_ANIMATION_MIN) + BLOB_ANIMATION_MIN;
		
		OreBlob blob = new OreBlob(name, pt,ImageStore.getImages(i_store, "blob"), temp * BLOB_ANIMATION_RATE_SCALE, rate);
		blob.scheduleEntity(this,ticks,i_store);
		return blob;
	}
	
	public Quake createQuake(Point pt, long ticks, HashMap<String, ArrayList<PImage>> iStore)
	{
		Quake quake = new Quake("quake", pt, ImageStore.getImages(iStore, "quake"), QUAKE_ANIMATION_RATE);
		quake.scheduleEntity(this,ticks, iStore);
		return quake;
	}
	
	public Point blobNextPosition(Point ePt, Point destPt)
	{
		int horz = Entities.sign(destPt.getX() - ePt.getX());
		Point newPt = new Point(ePt.getX() + horz, ePt.getY());
		if(horz == 0 || (isOccupied(newPt) && !(getTileOccupant(newPt) instanceof Ore)))
		{
			int vert = Entities.sign(destPt.getY() - ePt.getY());
			newPt = new Point(ePt.getX(),ePt.getY()+ vert);
			
			if (vert == 0 || (isOccupied(newPt) && !(getTileOccupant(newPt) instanceof Ore)))
			{
				newPt = new Point(ePt.getX(), ePt.getY());
			}
		}
		return newPt;
	}
	
	public int getNum_rows() {
		return num_rows;
	}
	public void setNum_rows(int num_rows) {
		this.num_rows = num_rows;
	}
	public int getNum_cols() {
		return num_cols;
	}
	public void setNum_cols(int num_cols) {
		this.num_cols = num_cols;
	}
	
	/***
	 * finds the entity of a certain type that is closest to a point in the world
	 * @param indexes, list of indexes of entities of that certain type
	 * @param pt, point you're comparing to
	 * @return the entity closest to that point
	 */
	public Entities nearestEntity(ArrayList<Integer> indexes, Point pt)
	{
		if(indexes.size() > 0)
		{
			int near = indexes.get(0);
			for( int i = 0; i < indexes.size(); i++)
			{
				double temp = this.entity.get(near).getPosition().distance(pt);
				double temp2 = this.entity.get(i).getPosition().distance(pt);
				
				if(temp2 < temp)
				{
					near = i;
				}
						
			}
			return this.entity.get(near);
		}
		return null;
	}
	
	
}
